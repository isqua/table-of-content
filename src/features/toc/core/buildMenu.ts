import type { MenuItem, PageDescriptor, PageId, PageURL, TableOfContent } from '../types'

type BuildMenuBaseOptions = {
    url: PageURL
}

type BuildMenuTopLevelOptions = BuildMenuBaseOptions

type BuildMenuNestingOptions = BuildMenuBaseOptions & {
    parentId: string
    level: number
}

type BuildMenuOptions = BuildMenuTopLevelOptions | BuildMenuNestingOptions

type PageProps = {
    isActive: boolean
    level: number
}

const mapPageToMenuItem = (page: PageDescriptor, props: PageProps): MenuItem => ({
    id: page.id,
    parentId: page.parentId,
    title: page.title,
    url: page.url ?? '',
    isActive: props.isActive,
    level: props.level,
    hasChildren: Boolean(page.pages?.length),
})

const getPagesForParent = (toc: TableOfContent, parentId?: PageId): PageId[] => {
    if (!parentId) {
        return toc.topLevelIds
    }

    if (!(parentId in toc.entities.pages)) {
        return []
    }

    return toc.entities.pages[parentId]?.pages ?? []
}

export const buildMenu = (toc: TableOfContent, options: BuildMenuOptions): MenuItem[] => {
    const { url, parentId, level = 0 } = options as BuildMenuNestingOptions
    const menu: MenuItem[] = []
    const pagesIds = getPagesForParent(toc, parentId)

    for (const pageId of pagesIds) {
        const page = toc.entities.pages[pageId]

        if (page) {
            menu.push(mapPageToMenuItem(page, {
                isActive: page.url === url,
                level,
            }))
        }
    }

    return menu
}
