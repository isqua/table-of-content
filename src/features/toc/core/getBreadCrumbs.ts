import type { PageDescriptor, PageURL, PageId, TableOfContent } from '../types'
import { getCurrentPage } from './getCurrentPage'

/**
 * @param toc The TOC data
 * @param url Page URL
 * @returns Current page and the array of its ancestors from top to bottom
 */
export const getBreadCrumbs = (toc: TableOfContent, url?: PageURL): PageDescriptor[] => {
    const breadcrumbs: PageDescriptor[] = []

    // We trust the API, but not completely
    const visitedIds = new Set<PageId>()

    if (!url) {
        return breadcrumbs
    }

    let currentPage: PageDescriptor | undefined = getCurrentPage(toc, url)

    while (currentPage && toc.entities.pages[currentPage.id] && !visitedIds.has(currentPage.id)) {
        breadcrumbs.push(currentPage)
        visitedIds.add(currentPage.id)

        currentPage = toc.entities.pages[currentPage.parentId]
    }

    return breadcrumbs.reverse()
}
