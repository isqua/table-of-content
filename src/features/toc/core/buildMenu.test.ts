import { describe, expect, it } from 'vitest'

import type { TableOfContent } from '../types'
import { buildMenu } from './buildMenu'
import { getBreadCrumbs } from './getBreadCrumbs'

import tocFlat from '../../../test/fixtures/toc/flat.json'
import tocTwoLevels from '../../../test/fixtures/toc/two-levels.json'

describe('toc/buildMenu', () => {
    it('should build a menu and highlight current page', () => {
        const toc: TableOfContent = tocFlat
        const currentUrl = 'bar.html'
        const breadcrumbs = getBreadCrumbs(toc, currentUrl)
        const menu = buildMenu(toc, { url: currentUrl, breadcrumbs })

        expect(menu).toMatchSnapshot()
    })

    it('should build a menu even if there are no current page', () => {
        const toc: TableOfContent = tocFlat
        const currentUrl = ''
        const breadcrumbs = getBreadCrumbs(toc, currentUrl)
        const menu = buildMenu(toc, { url: currentUrl, breadcrumbs })

        expect(menu).toMatchSnapshot()
    })

    it('should build an empty menu if there are no top level ids ', () => {
        const toc: TableOfContent = {
            entities: { pages: {} },
            topLevelIds: [],
        }
        const currentUrl = ''
        const breadcrumbs = getBreadCrumbs(toc, currentUrl)
        const menu = buildMenu(toc, { url: currentUrl, breadcrumbs })

        expect(menu).toMatchSnapshot()
    })

    it('should show page children', () => {
        const toc: TableOfContent = tocTwoLevels
        const currentUrl = 'bar.html'
        const breadcrumbs = getBreadCrumbs(toc, currentUrl)
        const menu = buildMenu(toc, { url: currentUrl, breadcrumbs })

        expect(menu).toMatchSnapshot()
    })

    it('should build a nested menu', () => {
        const toc: TableOfContent = tocTwoLevels
        const currentUrl = 'bar-features.html'
        const breadcrumbs = getBreadCrumbs(toc, currentUrl)
        const menu = buildMenu(toc, { url: currentUrl, parentId: 'bar', breadcrumbs })

        expect(menu).toMatchSnapshot()
    })
})
