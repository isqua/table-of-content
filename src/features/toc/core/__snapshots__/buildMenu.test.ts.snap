// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`toc/buildMenu > should build a menu and highlight current page 1`] = `
[
  {
    "defaultOpenState": false,
    "hasChildren": false,
    "highlight": "active",
    "id": "bar",
    "level": 0,
    "parentId": "root",
    "title": "Bar",
    "url": "bar.html",
  },
  {
    "defaultOpenState": false,
    "hasChildren": false,
    "highlight": undefined,
    "id": "baz",
    "level": 0,
    "parentId": "root",
    "title": "Baz",
    "url": "baz.html",
  },
  {
    "defaultOpenState": false,
    "hasChildren": false,
    "highlight": undefined,
    "id": "foo",
    "level": 0,
    "parentId": "root",
    "title": "Foo",
    "url": "foo.html",
  },
]
`;

exports[`toc/buildMenu > should build a menu even if there are no current page 1`] = `
[
  {
    "defaultOpenState": false,
    "hasChildren": false,
    "highlight": undefined,
    "id": "bar",
    "level": 0,
    "parentId": "root",
    "title": "Bar",
    "url": "bar.html",
  },
  {
    "defaultOpenState": false,
    "hasChildren": false,
    "highlight": undefined,
    "id": "baz",
    "level": 0,
    "parentId": "root",
    "title": "Baz",
    "url": "baz.html",
  },
  {
    "defaultOpenState": false,
    "hasChildren": false,
    "highlight": undefined,
    "id": "foo",
    "level": 0,
    "parentId": "root",
    "title": "Foo",
    "url": "foo.html",
  },
]
`;

exports[`toc/buildMenu > should build a nested menu 1`] = `
[
  {
    "defaultOpenState": false,
    "hasChildren": false,
    "highlight": undefined,
    "id": "bar_install",
    "level": 0,
    "parentId": "bar",
    "title": "Bar: Install",
    "url": "bar-install.html",
  },
  {
    "defaultOpenState": false,
    "hasChildren": false,
    "highlight": "active",
    "id": "bar_features",
    "level": 0,
    "parentId": "bar",
    "title": "Bar: Features",
    "url": "bar-features.html",
  },
]
`;

exports[`toc/buildMenu > should build an empty menu if there are no top level ids  1`] = `[]`;

exports[`toc/buildMenu > should show page children 1`] = `
[
  {
    "defaultOpenState": true,
    "hasChildren": true,
    "highlight": "active",
    "id": "bar",
    "level": 0,
    "parentId": "root",
    "title": "Bar",
    "url": "bar.html",
  },
  {
    "defaultOpenState": false,
    "hasChildren": false,
    "highlight": undefined,
    "id": "baz",
    "level": 0,
    "parentId": "root",
    "title": "Baz",
    "url": "baz.html",
  },
  {
    "defaultOpenState": false,
    "hasChildren": false,
    "highlight": undefined,
    "id": "foo",
    "level": 0,
    "parentId": "root",
    "title": "Foo",
    "url": "foo.html",
  },
]
`;
